Lib/ufo2ft/filters/unnestComponents.py

```diff
@@ -0,0 +1,62 @@
+from fontTools.misc.transform import Transform
+from ufo2ft.filters import BaseFilter
+
+import logging
+
+
+logger = logging.getLogger(__name__)
+
+
+class UnnestComponentsFilter(BaseFilter):
+
+    def set_context(self, font, glyphSet):
+        ctx = super(UnnestComponentsFilter, self).set_context(font, glyphSet)
+
+        ctx.stats = {'unnested': 0}
+
+        return ctx
+
+    def __call__(self, font, glyphSet=None):
+        if super(UnnestComponentsFilter, self).__call__(font, glyphSet):
+            stats = self.context.stats
+            if stats['unnested']:
+                logger.info('Unnested composite glyphs: %i' %
+                            stats['unnested'])
+
+    def filter(self, glyph):
+        if not glyph.components:
+            return False
+        unnested = False
+        for comp in glyph.components:
+            unnested_tuples = _unnestComponent(comp)
+            if len(unnested_tuples) == 1:
+                if unnested_tuples[0] == (comp.baseGlyph, comp.transformation):
+                    continue
+            unnested = True
+            index = glyph.componentIndex(comp)
+            glyph.removeComponent(comp)
+            for unnested in reversed(unnested_tuples):
+                base_comp = glyph.instantiateComponent()
```

Comment:
maybe this `instantiateComponent` method is too specific to defcon's API. I see for example fontParts does not have this.
Does it not work if you insert a generic `defcon.Component`?

Or maybe wouldn't it be better (more portable) if you used the pen API to draw the components onto the glyph object? That is, you get the pen (or point pen) associated with the glyph and call the addComponent method.

I guess in this case the new "unnested" components will all be drawn at the end of the glyphs.components list, whereas here you're using insertComponent (instead of appendComponent) to try to keep the order?
Not sure it's even worth to keep the order.

Reply:
I'd still prefer to use a generic pen.addComponent instead of relying on defcon-specific `instantiateComponent` and `insertComponent` methods